keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -32,218 +31,110 @@ status: provisional
-
-#### End-User Behavior
-
-Add a new `--trace` argument to `kubectl`, which generates a new trace context, sets the trace context to be sampled, attaches the context to all modified objects, and uses the context when sending requests to the API Server.  The option is disabled by default.  Note that by attaching a trace context to the initial object creation, this will cause all object modification done by controllers to propagate the context through to all changes made by the system that are driven by the initial user action.
-
-Add `context.Context` arguments to k8s.io/client-go client functions.  This will allow users and components to associate API calls with the context of the involved object.  In some cases, such as object creation, we can automatically attach the SpanContext of the provided context to the created object, making propagation simpler.
+### Tracing API Requests
 
-This also enables kubernetes to be a composable part of a larger system. For example, if an end-user's service creates a pod as part of handing a request, it could do:
-```golang
-ctx, span := trace.StartSpan(preexistingContext, “create-my-pod”)
-defer span.End()
-pod, err := c.CoreV1().Pod(myPod.Namespace).Create(ctx, myPod)
-if err != nil {
-    return err
-}
-waitForPodToBeRunning(ctx, myPod)
-return nil
-```
+We will wrap the API Server's http server and http clients with [othttp](https://github.com/open-telemetry/opentelemetry-go/tree/master/plugin/othttp) to get spans for incoming and outgoing http requests, and add the [otgrpc](https://github.com/grpc-ecosystem/grpc-opentracing/tree/master/go/otgrpc) DialOption to the etcd grpc client.  This generates spans for all sampled incoming requests and propagates context with all client requests.
 
-A previous iteration of this proposal suggested controllers should export a "Root Span" when ending a trace (described in [Controller Behavior](#controller-behavior) above).  However, that would limit a trace to being associated with a single object, since a "Root Span" defines the scope of the trace.  More generally, we shouldn't assume that the creation or update of a single object represents the entirety of end-user intent.  The user or system using kubernetes determines what the user intent is, not kubernetes controllers.
+Note that some clients of the API Server, such as webhooks, may make reentrant calls to the API Server.  To gain the full benefit of tracing, such clients should propagate context with requests back to the API Server.
 
-Tracing in a kubernetes cluster must be a composable component within a larger system, and allow external users or systems to define the "Root Span" that defines and bounds the scope of a trace.
+### Kubectl integration
```

Comment:
This feels a little odd to me. Typically sampling decisions are not made by the client, but rather based on the latency percentile the request happens to end up in. This also seems like a possible attack vector by putting load on the backing tracing storage through clients.

Reply:
That is different from my understanding of the typical setup, at least for the in-process sampling decision.  The [ProbabilitySampler](https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/trace/sampling.go#L84), is the "normal" built-in sampler.  ProbabilitySampler preserves the sampling decision of the parent context.  That ensures the entire trace is sampled and we don't end up with fragments if components make sampling decisions independently.

Tail-based sampling is pretty cool, and does what you are suggesting (sample based on the latency percentile).  If we wanted to, we could add configuration to the API Server to enable running with AlwaysSample so that the sampling decision can be made at the aggregation layer, instead of the collection layer.  The kubectl integration here wouldn't matter in that case, as the trace might get dropped at the aggregation layer.

I agree with the concern about telemetry being an attack vector.  We definitely should ensure it can't DOS the API Server.  However, I don't think managing load on storage backends is within the scope of kubernetes.  That being said, if you are using the OT Collector, you could use the [tail sampling processor](https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor#tail-sampling-processor) to rate limit.  If we choose to attach attributes such as namespace to spans, that would enable different policies for different namespaces.

Reply:
makes sense, I guess what I'm saying is, let's essentially assume that every request is always traced from a scalability perspective, sampling as you said can be done by the collector

Reply:
I'm also confused by this, is this really the primary way a trace would be used? I'd expect most of the things you'd want to trace would be controller calls?

Reply:
I don't think we care about controller calls in particular.  Without the kubectl integration, we would just have a probability sampler, which would be configurable through a command line flag.  It would randomly trace X% of incoming requests.  The other option, as described above, would be to do tail-based sampling to get all of the slowest (for example) requests.  But I don't consider that to be the common case.

My assumption is that it would be useful when debugging slow calls to be able to specify that a given call is sampled.  If I am trying to reproduce and debug a slow request, I don't want to have to make the same call 1000 times to ensure I get a trace out of it.  @logicalhan may have more specific examples of where this kind of functionality would have been useful in his experience.

That being said, I could exclude the kubectl integration given that enhancements freeze is tomorrow, and the sig-cli reviewer hasn't looked at this yet.

Reply:
I removed the kubectl integration portion for now.  We can always add it back later.